How to define spec.type for pluggable components. (#3155)
dapr/docs: daprdocs/content/en/operations/components/pluggable-components-registration.md

By default, Dapr looks for [Unix Domain Socket][uds] files in the folder in `/tmp/dapr-components-sockets`.

> Filenames in this folder are significant for component registration. They must be formed by appending the component's **name** with a file extension of your choice, more commonly `.sock`. For example, the filename `my-component.sock` is a valid UDS file name for a component named `my-component`.

Reviewer: ```suggestion
Filenames in this folder are significant for component registration. They must be formed by appending the component's **name** with a file extension of your choice, more commonly `.sock`. For example, the filename `my-component.sock` is a valid Unix Domain Socket file name for a component named `my-component`.
```

Since you are running Dapr in the same host as the component, verify this folder and the files within it are accessible and writable by both your component and Dapr.

> ### Building-blocks discovery from  behind the same component

Author: We could probably rename this for section to "building block discovery and multiplexing"

Reviewer 2: While I understand using "building block" to differentiate from "component", and could live with it, I still feel it's a bit imprecise.  To me, "component discovery and multiplexing" feels most accurate.

Author: I am happy with that :)

Just pushed this new section title.

> ### Building-blocks discovery from  behind the same component

Author: ```suggestion
### Component discovery and multiplexing
```

> A pluggable component accessible through a [Unix Domain Socket][uds] can host multiple distinct bulding blocks. During the components initial discovery process, Dapr will use reflection to enumerate all building blocks behind a UDS. The `my-component` pluggable component from the example above could contain both a state store (`state`) and a Pub/Sub (`pubsub`) building-blocks.

Reviewer 3: ```suggestion
A pluggable component accessible through a [Unix Domain Socket][UDS] can host multiple distinct component APIs . During the components' initial discovery process, Dapr uses reflection to enumerate all the component APIs behind a UDS. The `my-component` pluggable component in the example above can contain both state store (`state`) and a pub/sub (`pubsub`) component APIs.
```

> The most common use-case is for a 1:1 mapping between pluggable components and building blocks they expose. That said, having the possibility of consolidating multiple building blocks under the same pluggable components might be healpful, specially to ease DevOps burden at the expense of fault tolerance and security. Please weight the pros and cons and, if in doubt, stick to a 1:1 approach.

Reviewer: ```suggestion
Typically, a pluggable component implements a single component API for packaging and deployment. However, at the expense of having dependencies and security, a pluggable component can have multiple component APIs implemented to ease the deployment burden. Best practice for isolation, fault tolerance, and security is a single component API implementation for each pluggable component.
```

> The most common use-case is for a 1:1 mapping between pluggable components and building blocks they expose. That said, having the possibility of consolidating multiple building blocks under the same pluggable components might be healpful, specially to ease DevOps burden at the expense of fault tolerance and security. Please weight the pros and cons and, if in doubt, stick to a 1:1 approach.

Reviewer 3: ```suggestion
Typically a pluggable component implements a single component API for packaging and deployment. However a pluggable component can have multiple component APIs implemented  to ease the deployment burden at the expense of having dependencies and potentially security. The recommendation for isolation, fault tolerance and security is a single component API implementation for each pluggable components, however this is a guideline. 
```

Reviewer 2: Suggest "component API" instead of "component interface" to be consistent with above text.  (I don't have a strong preference for either one, but consistent phrasing would be best.)

> Typically, a pluggable component implements a single component API for packaging and deployment. However, at the expense of having dependencies and security, a pluggable component can have multiple component APIs implemented to ease the deployment burden. Best practice for isolation, fault tolerance, and security is a single component API implementation for each pluggable component.

Author: I think there is something off with the _",at the expense of having dependencies and security,"_. Don't we have _dependencies_ with a single API behind a pluggable component? _Security_ what? Problems?

How about 

```suggestion
Typically, a pluggable component implements a single component API for packaging and deployment. However, at the expense of increasing its dependencies and broadening its security attack surface, a pluggable component can have multiple component APIs implemented. This could be done to ease the deployment and monitoring burden. Best practice for isolation, fault tolerance, and security is a single component API implementation for each pluggable component.
```

## Define the component

Define your component using a [component spec]({{< ref component-schema.md >}}). Your component's `spec.type` value is made by concatenating the following 2 parts with a `.`:

> 1. the component's **building block type** (`state`, `pubsub` etc)

Reviewer 3: ```suggestion
1. The component's API (`Store`, `PubSub`, `Binding')
```

Author: Actually, the Component API or Interfaces types are `state`, `pubsub` and `bindings`. I cannot take this suggestion but I will fix this to something better.

Reviewer: ```suggestion
1. The component's API (`state`, `pubsub`, `bindings`, etc)
```

> 1. the component's **building block type** (`state`, `pubsub` etc)

Reviewer 2: Might we be more explicit and enumerate all the possible types?

Author: We only have 1 more (bindings)  for the time being but this would be another list to update as we add support to more building blocks/components. The "etc" was a deliberate choice to avoid this getting out of sync 😅

Reviewer 2: Yes, I know it's a maintenance burden, but sometimes having concrete examples (for confusing configuration) can be helpful rather than require user to interpolate.

Author: I can add a link to this table here https://docs.dapr.io/operations/components/pluggable-components/pluggable-components-overview/#implementing-a-pluggable-component.

To add to this confusiong: on that page it uses "Component API" for what I am using "building block" for.

Reviewer 3: Building Block is the wrong term here, which are these Dapr APIs https://docs.dapr.io/reference/api/
Pluggable components implement component APIs or interfaces. For example this https://github.com/dapr/components-contrib/blob/a494f3129d7c151131850cbfe72cf11e4379b95e/state/store.go#L24

> 2. its **name**, which is derived from the socket name, without the file extension. 

Reviewer: ```suggestion
2. The component's **name**, which is derived from the socket name, without the file extension. 
```

> You will need to define one [component spec]({{< ref component-schema.md >}}) for each building block exposed by your pluggable component's  [Unix Domain Socket][uds]. From the example above, the UDS `my-component.sock` from the previous example, exposes a pluggable component named `my-component` with both a `state` and a `pubsub` building blocks. Two components specs, each in their own YAML file placed in the resouces-path, will be required, one for `state.my-component` and another for `pubsub.my-component`.

Reviewer: ```suggestion
You will need to define one [component spec]({{< ref component-schema.md >}}) for each API exposed by your pluggable component's [Unix Domain Socket][uds]. The Unix Domain Socket `my-component.sock` from the previous example exposes a pluggable component named `my-component` with both a `state` and a `pubsub` API. Two components specs, each in their own YAML file, placed in the `resources-path`, will be required: one for `state.my-component` and another for `pubsub.my-component`.
```

> For instance, the component spec for `state.my-component` could be as follow

Reviewer 2: ```suggestion
For instance, the component spec for `state.my-component` could be:
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: my-production-state-store
spec:
  type: state.my-component
  version: v1
  metadata:
```

> In this sample above, notice the following:

Reviewer: ```suggestion
In the sample above, notice the following:
```

> * The contents of the field `spec.type` is `state.my-component`: this refers to a State Store being exposed as a pluggable component named `my-component`.

Reviewer: ```suggestion
* The contents of the field `spec.type` is `state.my-component`, referring to a state store being exposed as a pluggable component named `my-component`.
```

> * The field `metadata.name`, which is the name of the state store beging defined here, is not related with the pluggable component name.

Reviewer: ```suggestion
* The field `metadata.name`, which is the name of the state store being defined here, is not related to the pluggable component name.
```
